Update structured concurrency proposal to reflect latest changes to the Task API - @Sendable, etc.

nikitamounier opened the pull request:
This commit offers to change some of the code blocks which still had some of the older and deprecated features like `@concurrent`. More generally, though, I changed the function signatures so that they reflect how they are currently in the Task.swift, TaskGroup.swift and TaskCancellation.swift files, and also some of the implementations of the static properties like `currentPriority` which now use the `withUnsafeCurrentTask` function. Also, `Task.Handle` doesn't seem to have the `isCancelled` property anymore, which is now reflected in the proposal.

For the "free" (excuse my lack of a better term) functions like `detach` and `withTaskGroup` which were previously static, I chose to explicitly make them "public" in the document to emphasise that these are to be used "in the wild".

Now, this is truly a nitpick, but I feel like calling static properties like `isCancelled` "functions" is unnatural and uncommon, which is why I changed them to "properties". However, if you have your own reasons for this, I completely understand, and would be happy to revert those changes.

ktoso commented on proposals/0304-structured-concurrency.md:
```diff
@@ -583,13 +583,10 @@ Task handles also provide the ability to cancel a task programmatically:
   func cancel()
-  
-  /// Determine whether the task was cancelled.
-  var isCancelled: Bool { get }
```
This is supposed to be here

ktoso commented on proposals/0304-structured-concurrency.md:
````diff
@@ -583,13 +583,10 @@ Task handles also provide the ability to cancel a task programmatically:
 }
 ```
 
-As noted elsewhere, cancellation is cooperative: the task will note that it has been cancelled and can choose to return earlier (either via a normal return or a thrown error, as appropriate). `isCancelled` can be used to determine whether a particular task was ever cancelled.
````
This should stay

ktoso commented on proposals/0304-structured-concurrency.md:
````diff
@@ -600,7 +597,7 @@ extension Task.Handle {
 }
 ```
 
-Getting the handle's task allows us to check if the work we're about to wait on perhaps was already cancelled (by calling `handle.isCancelled`), or query at what priority the task is executing.
````
this should stay

ktoso commented on proposals/0304-structured-concurrency.md:
```diff
@@ -609,17 +606,17 @@ A new, detached task can be created with the `detach` operation. The resulting t
 
 /// Create a new, detached task that produces a value of type `T` or throws an error.
 @discardableResult
-static func detach<T: Sendable>(
```
thanks for removing the `static`, we don't usually do `public` for everything in proposals -- every single function would have it. I would prefer not adding them, thanks

ktoso commented on proposals/0304-structured-concurrency.md:
```diff
@@ -609,17 +606,17 @@ A new, detached task can be created with the `detach` operation. The resulting t
+public func detach<T, Failure>(
   priority: Task.Priority = .unspecified,
-  operation: @escaping @concurrent () async throws -> T
-) -> Task.Handle<T, Error>
```
here the proposal is correct and the impl is in the wrong; thanks for reminding me though!

ktoso commented on proposals/0304-structured-concurrency.md:
```diff
@@ -738,7 +735,7 @@ extension Task {
 #### Cancellation
 
-It is possible to query for cancellation from within a synchronous task, e.g. while iterating over a loop and wanting to check if we should abort its execution by using the static `Task.isCancelled` function:
+It is possible to query for cancellation from within a synchronous task, e.g. while iterating over a loop and wanting to check if we should abort its execution by using the static `Task.isCancelled` property:
```
thanks 👍

ktoso commented on proposals/0304-structured-concurrency.md:
````diff
@@ -763,30 +760,34 @@ extension Task {
 ```swift
 extension Task {
   static var isCancelled: Bool { 
-    Task.current?.isCancelled ?? false
````
Another case where the proposal is more up to date than the impl, please keep it.

We don't need to drop to unsafe to implement this.

ktoso commented on proposals/0304-structured-concurrency.md:
````diff
@@ -763,30 +760,34 @@ extension Task {
 ```swift
 extension Task { 
   static var currentPriority: Task.Priority { 
-    Task.current?.priority ?? Task.Priority.default
````
Please don't change this; the proposal is cleaner than the implementation and we don't need to show unsafe here.

ktoso commented on proposals/0304-structured-concurrency.md:
```diff
@@ -1018,7 +1019,7 @@ extension ThrowingTaskGroup: AsyncSequence {
   /// Wait for a task to complete and return the result or thrown error packaged in
   /// a `Result` instance. Returns `nil` only when there are no tasks left in the group.
-  mutating func nextResult() async -> Result<ChildTaskResult, Error>?
+  mutating func nextResult() async throws -> Result<ChildTaskResult, Error>?
```
This is wrong. Thanks for helping spot this though, will add a test covering it.

We either return nil, success or failure, there is no room for throwing here; the signature in the proposal is correct.

ktoso commented on proposals/0304-structured-concurrency.md:
```diff
@@ -905,9 +906,9 @@ static func withTaskGroup<ChildTaskResult: Sendable, GroupResult>(
 /// - if the body throws:
 ///   - all tasks remaining in the group will be automatically cancelled.
-static func withThrowingTaskGroup<ChildTaskResult: Sendable, GroupResult>(
+public func withThrowingTaskGroup<ChildTaskResult: Sendable, GroupResult>(
```
Thanks for removing the static, let's not add public to things though -- it is done inconsistently, say all spawn et all are public too of course but the proposal text misses that.

Let's please not  add public in the proposal snippets.

ktoso commented on proposals/0304-structured-concurrency.md:
```diff
@@ -421,8 +421,8 @@ As already shown in the above examples, it is possible to call functions that in
-static func withUnsafeCurrentTask<T>(
-    operation: (UnsafeCurrentTask?) throws -> T
+public func withUnsafeCurrentTask<T>(
+    body: (UnsafeCurrentTask?) throws -> T
```
operation is nicer; the proposal is ahead of impl and better a bit, please don't change this

ktoso commented on proposals/0304-structured-concurrency.md:
````diff
@@ -609,17 +606,17 @@ A new, detached task can be created with the `detach` operation. The resulting t
 ```swift
 /// Create a new, detached task that produces a value of type `T`.
 @discardableResult
-static func detach<T: Sendable>(
````
I'll actually confirm what should be done here... this was in flux still recently

ktoso replied:
Okey this turned out to be tricky, the change is good though -- no Sendable on the `T` BUT we force it on the `Task.Handle<T, ...>`

ktoso commented on proposals/0304-structured-concurrency.md:
```diff
@@ -609,17 +606,17 @@ A new, detached task can be created with the `detach` operation. The resulting t
 /// Create a new, detached task that produces a value of type `T` or throws an error.
 @discardableResult
-static func detach<T: Sendable>(
+public func detach<T, Failure>(
```
please no public and no Failure param; the impl does this wrong; the error must be hardcoded to `Error`

ktoso commented on proposals/0304-structured-concurrency.md:
```diff
@@ -655,7 +652,7 @@ extension Task {
   }
 
-  func checkCancellation() throws
+  static func checkCancellation() throws
```
thank you 👍

ktoso reviewed (changes requested):
Thanks for the checks here! This uncovered a few things the impl is lagging behind on.

Please remove any unnecessary changes I commented on -- it really isn't much that needs fixing; mostly the concurrent -> Sendable and static removals.

Thank you for those though! Very helpful to have another skim of this comparing with the impl.

nh7a commented:
Can you also add:
```swift
    for await finishedStep in group {
      switch finishedStep {  // <=== typo
        case .veggies(let v): veggies = v
        case .meat(let m): meat = m
        case .oven(let o): oven = o
      }
    }
```

nikitamounier commented:
Alright, I reverted all the unnecessary changes, and fixed that typo! Let me know whether the generic type `T` needs to be constrained to `Sendable` in the `detach` function.

ktoso commented:
> Let me know whether the generic type T needs to be constrained to Sendable in the detach function.

This one I had to check about future plans with sendability checking with the team, turns out that indeed we don't put the requirement on the T on detach, BUT we do put it on the handle's T :-)

ktoso commented on proposals/0304-structured-concurrency.md:
````diff
@@ -421,8 +421,8 @@ As already shown in the above examples, it is possible to call functions that in
 ```swift
-static func withUnsafeCurrentTask<T>(
-    operation: (UnsafeCurrentTask?) throws -> T
+public func withUnsafeCurrentTask<T>(
````
```suggestion
func withUnsafeCurrentTask<T>(
```

ktoso reviewed (approved):
Excellent, thanks for the followups 👍

nikitamounier commented:
Great! Pleasure working with you.

nikitamounier commented:
I missed one last thing - the `withTaskCancellationHandler` shouldn't be static on line 679.

ktoso commented:
Thanks fixed https://github.com/apple/swift-evolution/pull/1324
